compiler/rustc_middle/src/ty/util.rs

```diff
@@ -1544,12 +1544,12 @@ pub fn intrinsic(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<Symbol> {
     {
         Some(tcx.item_name(def_id.into()))
     } else if let Some(name) = tcx.get_attr(def_id, sym::rustc_intrinsic) {
-        Some(
+        let name =
             name.get_normal_item().meta(rustc_span::DUMMY_SP).unwrap().meta_item_list().unwrap()[0]
                 .ident()
                 .unwrap()
-                .name,
-        )
+                .name;
+        tcx.intrinsics_implemented_by_backend(()).contains(&name).then_some(name)
```

Comment:
This seems like it is bound to go out of sync with what the backend actually supports. Would it be possible to instead obtain the fallback MIR body inside the codegen backend when the backend determines that there is no implementation? Also there are various places where `is_intrinsic` doesn't just affect codegen for the local crate. For example the lower intrinsics MIR pass (which uses `is_intrinsic`) affects the MIR bodies recorded in the crate metadata, const eval will use `is_intrinsic` to skip counting any execution cost for intrinsics. The value of `is_intrinsic` is also recorded in the crate metadata and thus only the codegen backend with which libcore is compiled effectively determines if `is_intrinsic` would return `true`. This completely breaks when mixing codegen backends, like mixing cg_clif with a cg_llvm sysroot as using the rustc-codegen-cranelift-preview component would do.

Reply:
Hmm... I wanted to avoid that, but I agree it's annoying that they go out of sync.

I'll think about how to make backends able to call the default body.

Reply:
Done. The backends need to implement this logic themselves every time (I implemented it for cranelift), but that indeed seems like the best solution.